cranelift/codegen/src/isa/x64/lower.isle

```diff
@@ -3491,23 +3491,150 @@
 
 ;; Rules for `udiv` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
-(rule (lower (udiv a @ (value_type ty) b))
-      (div_or_rem (DivOrRemKind.UnsignedDiv) a b))
+;; The inputs to the `div` instruction are different for 8-bit division so
+;; it needs a special case here since the instruction being crafted has a
+;; different shape.
+(rule 2 (lower (udiv a @ (value_type $I8) b))
+        (x64_div8 (extend_to_gpr a $I32 (ExtendKind.Zero))
+                  (nonzero_divisor $I8 b)
+                  $false))
+
+;; 16-to-64-bit division is all done with a similar instruction and the only
+;; tricky requirement here is that when div traps are disallowed the divisor
+;; must not be zero.
+(rule 1 (lower (udiv a @ (value_type (fits_in_64 ty)) b))
+        (x64_div_quotient a
+                          (imm $I64 0)
+                          (nonzero_divisor ty b)
+                          (raw_operand_size_of_type ty)
+                          $false))
+
+;; Helper to place `Value` into a `Gpr` while possibly trapping if it's zero.
+;;
+;; If the `avoid_div_traps=true` codegen setting is specified then the value
+;; is checked for zero and a trap happens before the value is returned as a
+;; register here.
+(decl nonzero_divisor (Type Value) Gpr)
+
+;; As a special-case if the divisor is a constant number which is nonzero then
+;; no matter what there's no checks necessary.
+(rule 2 (nonzero_divisor ty (iconst (u64_from_imm64 (u64_nonzero n))))
+        (imm ty n))
+
+;; No checks necessary when `avoid_div_traps=false`
+(rule 1 (nonzero_divisor ty val)
+        (if-let $false (avoid_div_traps))
+        val)
+
+;; Base case executes a `test` instruction and trap if it's equal to itself.
```

Comment:
Slightly unclear comment here -- not quite "equal to itself" but "zero flag set" (Z is only eq for cmp). Maybe just "and trap if zero"?